File: Source/ch/nolix/primitive/container/List.java
//package declaration
package ch.nolix.primitive.container;

//Java import
import java.util.Iterator;

//class
public final class List<E> implements Iterable<E> {
	
	//optional attributes
	private ListNode<E> beginNode;
	private ListNode<E> endNode;
	
	//constructor
	public List() {}
	
	//constructor
	public List(final E element) {
		addAtBegin(element);
	}
	
	//method
	public void addAtBegin(final E element) {
		
		final ListNode<E> node = new ListNode<E>(element);
		
		if (isEmpty()) {
			beginNode = node;
			endNode = node;
		}
		else {
			node.setNextNode(beginNode);
			beginNode = node;
		}
	}
	
	//method
	public void addAtEnd(final E element) {
		
		final ListNode<E> node = new ListNode<E>(element);
		
		if (isEmpty()) {
			beginNode = node;
			endNode = node;
		}
		else {
			endNode.setNextNode(node);
			endNode = node;
		}
	}
	
	//method
	public E getRefFirst() {
		
		supposeIsNotEmpty();
		
		return beginNode.getRefElement();
	}

	//method
	public boolean isEmpty() {
		return (beginNode == null);
	}
	
	//method
	public void removeFirst() {
		
		supposeIsNotEmpty();
		
		if (!beginNode.hasNextNode()) {
			beginNode = null;
			endNode = null;
		}
		else {
			beginNode = beginNode.getRefNextNodeOrNull();
		}
	}

	//method
	/**
	 * @return a new {@link Iterator} for the current {@link List}.
	 */
	public Iterator<E> iterator() {
		return new ListIterator<E>(beginNode);
	}
	
	//method
	private void supposeIsNotEmpty() {
		if (isEmpty()) {
			throw new RuntimeException("The current list is empty.");
		}
	}
}

File: TypeScriptSource/Nolix/Core/Container/List.ts
//reference
/// <reference path="ListNode.ts" />

//namespace declaration
namespace Nolix.Core.Container {

    //cass
    export class List<E> {

        //attribute
        private elementCount: number = 0;

        //optional attributes
        private beginNode: ListNode<E>;
        private endNode: ListNode<E>;

        //method
        public addAtBegin(element: E): List<E> {

            const previousNode = new ListNode<E>(element);

            if (this.isEmpty) {
                this.beginNode = previousNode;
                this.endNode = previousNode;
            }
            else {
                previousNode.setNextNode(this.beginNode);
                this.beginNode = previousNode;
            }

            this.elementCount++;

            return this;
        }

        //method
        public addAtEnd(element: E): List<E> {

            const nextNode = new ListNode<E>(element);

            if (this.isEmpty) {
                this.beginNode = nextNode;
                this.endNode = nextNode;
            }
            else {
                this.endNode.setNextNode(nextNode);
                this.endNode = nextNode;
            }

            this.elementCount++;

            return this;
        }

        //method
        public containsAny(): boolean {

            //For a better performance, this implementation does not use all comfortable methods.
            return (this.elementCount > 0);
        }

        //method
        public getRefFirst(): E {
            
            this.supposeIsNotEmpty();

            return this.beginNode.getRefElement();
        }

        //method
        public getRefLast(): E {

            this.supposeIsNotEmpty();

            return this.endNode.getRefElement();
        }

        //method
        public getSize(): number {
            return this.elementCount;
        }

        //method
        public isEmpty(): boolean {

            //For a better performance, this implementation does not use all comfortable methods.
            return (this.elementCount == 0);
        }

        //method
        private supposeIsNotEmpty(): void {
            if (this.isEmpty) {
                throw new Error("The current list is empty.");
            }
        }
    }
}
